### ISSUE TITLE
Connecting to new app throws stream exception

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/devtools#3302]
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: Bad state: Already cancelled
#0      StreamQueue._checkNotClosed (package:async/src/stream_queue.dart:521:20)
#1      StreamQueue.cancel (package:async/src/stream_queue.dart:390:5)
#2      DdsExtension.onEventWithHistory.<anonymous closure> (package:dds/vm_service_extensions.dart:94:20)
#3      _rootRun (dart:async/zone.dart:1420:47)
#4      _CustomZone.run (dart:async/zone.dart:1328:19)
#5      _FutureListener.handleWhenComplete (dart:async/future_impl.dart:203:18)
#6      Future._propagateToListeners.handleWhenCompleteCallback (dart:async/future_impl.dart:737:39)
#7      Future._propagateToListeners (dart:async/future_impl.dart:793:11)
#8      Future._addListener.<anonymous closure> (dart:async/future_impl.dart:466:9)
#9      _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#10     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
```

CC @bkonyi - is there a certain way we are supposed to be handling / cancelling calls to onEventWithHistory?

**Comment by bkonyi** [comment]
Not that I know of. Is it possible that you're cancelling the returned stream more than once?

**Comment by kenzieschmoll** [comment]
We listen to the same stream in multiple places in DevTools - does each call to `onEventWithHistory` return the same stream or a new instance?

**Comment by kenzieschmoll** [comment]
In DevTools we call `.cancel()` on the `StreamSubscription` object we create with `stream.listen` - we never actually cancel the stream itself.

**Comment by kenzieschmoll** [comment]
This is the leading crash right now. 30k+ exceptions in the last 7 days.

**Comment by bkonyi** [comment]
Can you reproduce locally?

**Comment by kenzieschmoll** [comment]
Yeah. Repro steps are to connect DevTools to a flutter app, then connect to a new app (click device info in bottom right of DevTools -> connect to new app).

**Comment by kenzieschmoll** [comment]
We listen to the stream using our auto_dispose code in DevTools. We then cancel the stream subscription on app disconnect: https://github.com/flutter/devtools/blob/201e68353105254bd307dfc60e46a1ad5d13d8c0/packages/devtools_app/lib/src/auto_dispose.dart#L43. Not sure if this is the first call to cancel() or the second, but this is definitely one of the two.

**Comment by bkonyi** [comment]
Oh awesome! Thanks for the details. I'll take a closer look today or tomorrow.

**Comment by bkonyi** [comment]
Fixed in `package:dds` `2.1.3`. We just need to bump the version in DevTools now.
